stdlib/domain.ml

```diff
@@ -9,6 +9,62 @@ module Raw = struct
     = "caml_ml_domain_cpu_relax"
 end
 
+module DLS = struct
+
+  type dls_state = Obj.t array
+
+  let unique_value = Obj.repr (ref 0)
+
+  external get_dls_state : unit -> dls_state = "%dls_get"
+
+  external set_dls_state : dls_state -> unit =
+    "caml_domain_dls_set" [@@noalloc]
+
+  let create_dls () =
+    let st = Array.make 8 unique_value in
+    set_dls_state st
+
+  let _ = create_dls ()
+
+  type 'a key = int * (unit -> 'a)
+
+  let key_counter = Atomic.make 0
+
+  let new_key f =
+    let k = Atomic.fetch_and_add key_counter 1 in
+    (k, f)
+
+  (* If necessary, grow the current domain's local state array such that [idx]
+   * is a valid index in the array. *)
+  let maybe_grow idx =
+    let st = get_dls_state () in
+    let sz = Array.length st in
+    if idx < sz then st
+    else begin
+      let rec compute_new_size s =
+        if idx < s then s else compute_new_size (2 * s)
+      in
+      let new_sz = compute_new_size sz in
+      let new_st = Array.make new_sz unique_value in
+      Array.blit st 0 new_st 0 sz;
+      set_dls_state new_st;
+      new_st
+    end
+
+  let set (idx, _init) x =
```

Comment:
I'm not sure you are being defensive enough in your implementation of an array of `Obj.t`s. The potential issues with the `Obj.t` being a float need to be handled carefully. See `Obj_array` in Base for an example of what is needed: https://github.com/janestreet/base/blob/master/src/obj_array.ml.

Reply:
Thanks @lpw25! I read through Base's `Obj_array`. IIUC, since we initialise the array with a non-float value, the array will not have `Double_array_tag`. Hence, `get` should be ok [1]. For `set`, if the element is wrapped in `Sys.opaque_identity`, then this will prevent flambda from incorrectly inferring that the array has `Double_array_tag` [2]. Does that address your concerns? Am I missing anything else?

[1] https://github.com/janestreet/base/blob/master/src/obj_array.ml#L50-L53
[2] https://github.com/janestreet/base/blob/master/src/obj_array.ml#L65-L69

Reply:
I think that's right. The `set` part was the one I was concerned about.

Reply:
I think the current PR is fine as it is, but I wonder if we shouldn't consider a safe version of this code, using one of the usual "universal type" embeddings. If I understand correctly, the code is not performance-critical, it should be O(1) but we can afford an extra dynamic check on access.